plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java

```diff
@@ -16,96 +16,108 @@
-   *     error from the transport layer while sending/receiving a HTTP request/response.)
-   */
-  public Credential getActiveCredential(IShellProvider shellProvider) throws IOException {
-    Credential credential = getCachedActiveCredential();
-
-    if (credential == null) {
-      credential = logIn(shellProvider);
-
-      IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-      eclipseContext.set(STASH_OAUTH_CRED_KEY, credential);
-    }
-    return credential;
+  private static final SortedSet<String> OAUTH_SCOPES = Collections.unmodifiableSortedSet(
+      new TreeSet<>(Arrays.asList(
+          "email", //$NON-NLS-1$
+          "https://www.googleapis.com/auth/cloud-platform" //$NON-NLS-1$
+      )));
+
+  private GoogleLoginState loginState;
+  private AtomicBoolean loginInProgress;
+
+  @VisibleForTesting
+  GoogleLoginService(
+      OAuthDataStore dataStore, UiFacade uiFacade, LoggerFacade loggerFacade) {
+    loginState = new GoogleLoginState(
+        Constants.getOAuthClientId(), Constants.getOAuthClientSecret(), OAUTH_SCOPES,
+        dataStore, uiFacade, loggerFacade);
+    loginInProgress = new AtomicBoolean(false);
   }
 
-  /**
-   * Returns the credential of an active user (among multiple logged-in users). Unlike {@link
-   * #getActiveCredential}, this version does not involve login process or make API calls.
-   * Returns {@code null} if no credential has been cached.
-   *
-   * Safe to call from non-UI contexts.
-   */
-  public Credential getCachedActiveCredential() {
-    IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-    return (Credential) eclipseContext.get(STASH_OAUTH_CRED_KEY);
+  public GoogleLoginService() {
+    this(new TransientOAuthDataStore(PlatformUI.getWorkbench().getService(IEclipseContext.class)),
+        new LoginServiceUi(), new LoginServiceLogger());
   }
 
-  private Credential logIn(IShellProvider shellProvider) throws IOException {
-    GoogleLoginBrowser loginBrowser = new GoogleLoginBrowser(
-        shellProvider.getShell(), Constants.getOAuthClientId(), OAUTH_SCOPES);
-    if (loginBrowser.open() != GoogleLoginBrowser.OK) {
-      return null;
+  @Override
+  public Credential getActiveCredential() {
+    if (!loginInProgress.compareAndSet(false, true)) {
+      LoginServiceUi.showErrorDialogHelper(
+          Messages.LOGIN_ERROR_DIALOG_TITLE, Messages.LOGIN_ERROR_IN_PROGRESS);
     }
 
-    GoogleAuthorizationCodeTokenRequest authRequest = new GoogleAuthorizationCodeTokenRequest(
-        Utils.getDefaultTransport(), Utils.getDefaultJsonFactory(),
-        Constants.getOAuthClientId(), Constants.getOAuthClientSecret(),
-        loginBrowser.getAuthorizationCode(),
-        GoogleLoginBrowser.REDIRECT_URI);
-
-    return createCredential(authRequest.execute());
+    // TODO: holding a lock for a long period of time (especially when waiting for UI events)
```

Comment:
Are you suggesting making `getActiveCredential()` return a Future or making this portion use a Future? If the latter case, that won't really help anything, because (if I'm not mistaken) this method still needs to block on `Future.get()` while holding a lock while the Future computation (i.e., login through UI) is on-going.